BZ1346042

csutherl opened the pull request:
In an update of tomcat version 7 (and 8 for Fedora) tomcat no longer writes to catalina.out, but instead writes to catalina.$(date).log. This means that sosreport doesn't collect the correct tomcat logs anymore. In addition, there are other logs which tomcat writes that would be helpful for debugging issues (manager*.log, host-manager*.log, localhost*.log, etc). This update will capture all of the logs when enabling the tomcat.log option, much like the apache plugin's log option which sosreport already uses.

Signed-off-by: Coty Sutherland <csutherl@redhat.com>

bmr-cymru commented:
Please use a meaningful subject when filing bugs - sos is supported on a wide range of distributions now and "BZNumber" doesn't really give any useful information.

Just a brief description of the actual problem is all that is needed - someone from the team will pick the issue up and add any tags, labels or other tracking bits that the team uses.

bmr-cymru commented on sos/plugins/tomcat.py:
```diff
@@ -24,6 +24,10 @@ class Tomcat(Plugin, RedHatPlugin):
+    option_list = [
+        ("log", "gathers all tomcat logs", "slow", False)
+    ]
+
```
Nack. There is a global `all_logs` option that is wired to the `--all-logs` command line switch. All plugins should use this rather than a plugin-specific option.

You can look at other plugins in the current tree for examples or take a look at the [plugin options](https://github.com/sosreport/sos/wiki/Plugin-options) wiki entry.

bmr-cymru commented on sos/plugins/tomcat.py:
```diff
@@ -34,6 +38,11 @@ def setup(self):
         self.add_copy_spec_limit(log_glob, sizelimit=limit)
 
+        if self.get_option("log"):
+            self.add_copy_spec([
```
Couple of comments - an `add_copy_spec()` call for a single item doesn't need the item to be wrapped in a list (the API accepts either a single string or a list of strings), so just:

```python
        self.add_copy_spec("/var/log/tomcat*/*")
```

is fine.

More generally though I think this patch needs more work - if the catalina logs are now obsolete there is really no point in collecting them at all (at least on newer installations) so we would still end up collecting nothing if the user did not specify `--all-logs`: I think there needs to be some middle ground between effectively collecting nothing and collecting absolutely everything (via globs and with all size limits disabled).

csutherl commented:
OK. I incorporated the suggestion to use all_logs instead of a plugin specific option, in addition to catalina.out (because it was used up until rhel-7.2) I am grabbing the catalina.log from the day sosreport is executed which should get the actual logging from that day, and I noticed some security issues in the collection of the server.xml, so I addressed those as well. Feedback welcome :)

csutherl commented:
I fixed the pep8 failures and incorporated the tomcat7 and tomcat8 package addition.

bmr-cymru commented on sos/plugins/tomcat.py:
```diff
@@ -13,6 +13,7 @@
 # Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 
 from sos.plugins import Plugin, RedHatPlugin
+from datetime import datetime
```
Not really a review comment, more a note-to-self: we have a number of plugins now doing time based calculations (tomcat, sytemd, pacemaker) so I've filed  #838 to keep track of that (there's no need for that work to block this pull request however).

bmr-cymru commented on sos/plugins/tomcat.py:
```diff
@@ -22,19 +23,39 @@ class Tomcat(Plugin, RedHatPlugin):
+                r"\/etc\/tomcat.*\/server.xml",
+                r"%s=(\S*)" % attr,
+                r'%s="********"' % attr
+            )
```
Ack, I'll take a closer look at this with real data some time but this looks good. The path regex matches are applied per-plugin so you can shorten the regex a little if you like (I'm happy either way - some folks don't like all the `\/` in them).

bmr-cymru commented on sos/plugins/tomcat.py:
```diff
@@ -22,19 +23,39 @@ class Tomcat(Plugin, RedHatPlugin):
+            # prepend 0
+            today = datetime.date(datetime.now()).isoformat()
+            log_glob = "/var/log/tomcat*/catalina.%s.log" % today
+            self.add_copy_spec_limit(log_glob, sizelimit=limit)
```
Thanks - this looks good & is very similar to other plugins. I'd like to streamline some of this in the future (across all plugins) but I think this gives us a reasonable lot of data for the time being - I'd not realised how little we captured on recent versions - I'm surprised nobody complained before!

bmr-cymru commented:
:+1:

csutherl commented:
Awesome :D I know for a fact that for tomcat the support engineers always manually grab stuff, regardless of having a sosreport so that's likely why. There is a new guy that raised this as a concerned, so I figured I could easily address it. Thanks for looking over it!
